.github/workflows/python.yaml

```diff
@@ -31,17 +29,28 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Set up Python 3.11
-      uses: actions/setup-python@v3
+      uses: actions/setup-python@v4
       with:
         python-version: 3.11
+        cache: 'pip'
+        cache-dependency-path: |
+          dev-deps.lock
+          pyproject.toml
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
         if [ -f dev-deps.lock ]; then pip install -r dev-deps.lock; fi
         pip install -e . --no-deps
-    - name: Lint with pre-commit
-      run: |
-        pre-commit run -av
+    - name: Cache linter tool caches
+      uses: actions/cache@v3
+      with:
+        path: |
+          ~/.cache/ruff/
+          ~/.cache/mypy/
+        key: ${{ runner.os }}-${{ hashFiles('dev-deps.lock', 'pyproject.toml', '.pre-commit-config.yaml') }}-linter-cache
+    - uses: pre-commit/action@v3.0.0
+      with:
+        extra_args: --show-diff-on-failure
```

Comment:
The action already supplies `--show-diff-on-failure`, see https://github.com/pre-commit/action/blob/5f528da5c95691c4cf42ff76a4d10854b62cbb82/action.yml#L19